react/more-react-concepts/reducing-state/reducing-state.md

```diff
@@ -0,0 +1,94 @@
+### Introduction
+
+You have probably heard about reducers before. In this lesson we are going to
+learn _what_ they are, _when_ to use them and _how_ to use them in React with the `useReducer` hook.
+
+### Lesson Overview
+
+This section contains a general overview of topics that you will learn in this lesson.
+
+* What are reducers.
+* When to use reducers.
+* What is the useReducer hook.
+
+### What are reducers?
+
+Reducers are pure functions that take a _previous state_, an _action_ and return
+a _new state_.
+
+The action is an object with a _type_ property describing what the user did. This object can also contain
+any other properties that our reducer might need to produce the new state.
+
+We can define a reducer that updates the state of a counter like this:
+
+~~~js
+function reducer(state, action) {
+  switch (action.type) {
+      case 'incremented_count': {
+        return { count: state.count + 1 }
+        }
+      case 'decreased_count': {
+          return { count: state.count - 1 }
+        }
+      case 'set_count': {
+          return { count: action.value }
+        }
+      default: {
+          throw error('unkown action: ' + action.type);
+        }
+    }
+}
+~~~
+
+Remember, reducers are pure functions so we shouldn't mutate the state.
+
+In React reducers can be very useful when we have components that update their state in complex or many ways,
+allowing us to separate the state logic from the component itself.
+
+### The useReducer hook
+
+React allows us to use reducers in our components through a hook called `useReducer`. This hook takes a reducer function and
+an initial state as arguments, then returns an array with two values: the current state and a `dispatch` function.
+
+This `dispatch` function receives an **action** object as argument, which is passed to our reducer function and 
+the returned value from it is used to update the state.
+
+~~~js
+const [state, dispatch] = useReducer(reducer, { count: 0 });
+
+function handleClick() {
+  dispatch({ type: 'incremented_count' });
+}
+~~~
+
+Similarly to `useState`, the `dispatch` function _only updates the state in the next render_, and it uses
+`Object.is` to determine if the state has changed, if it hasn't, then it won't trigger a re-render.
```

Comment:
> `Object.is` to determine if the state has changed, if it hasn't, then it won't trigger a re-render.


Interesting, can you provide a reference link/resource for this?

Reply:
sure :) this is for the dispatch function: https://react.dev/reference/react/useReducer#setstate-caveats
this is for useState, or more precisely, the 'set' function, I think I should make that change: https://react.dev/reference/react/useState#setstate-caveats

Should I add one of these resources or both somewhere?